### PR TITLE
HLM-5167 HLM-5158 adding navigation and respective restrictions

### PR DESCRIPTION
**Pull request by siddhant-nawale-egov**
Adding navigation and respective restrictions

## Summary by CodeRabbit

- **New Features**
  - Introduced a dynamic icon display in buttons based on conditions.
  - Added new components like `ActionBar`, `Stepper`, and `Toast` to enhance the microplanning process for employees.
  - Implemented a new data storage solution for session management.
  - Enhanced navigation with `ArrowBack` and `ArrowForward` components.
  - Integrated `Link` component for improved guideline navigation.
- **Bug Fixes**
  - Refined data completeness checks and state management in the `RuleEngine` component.
- **UI Changes**
  - Updated modal heading styles and microplanning card filtering logic.
  - Replaced `TimelineMicroplan` with a more interactive `Stepper` component.
  - Adjusted `Toast` component styling and positioning.
- **Refactor**
  - Improved handling of data storage and retrieval.
- **Chores**
  - Updated CSS version in the public index file to `1.0.8-microplan`.

### CODE REVIEW

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 1]
```diff
@@ -544,7 +590,7 @@ const Upload = ({ MicroplanName = "default", campaignType = "SMC" }) => {
           RightButtonHandler={() => UploadFileClickHandler(true)}
           sections={sections}
           footerLeftButtonBody={<ButtonType1 text={t("ALREADY_HAVE_IT")} />}
-          footerRightButtonBody={<ButtonType2 text={t("DOWNLOAD_TEMPLATE")} />}
+          footerRightButtonBody={<ButtonType2 text={t("DOWNLOAD_TEMPLATE")} download={true} />}
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:593]: That download option is just for the icon, it doesn't actually download the file.

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 2]
```diff
@@ -612,18 +658,18 @@ const Upload = ({ MicroplanName = "default", campaignType = "SMC" }) => {
...
       {loaderActivation && <LoaderWithGap text={"FILE_UPLOADING"} />}
-      {toast && toast.state === "success" && <Toast label={toast.message} onClose={() => setToast(null)} />}
+      {toast && toast.state === "success" && <Toast style={{ bottom: "5.5rem" }} label={toast.message} onClose={() => setToast(null)} />}
       {toast && toast.state === "error" && (
-        <Toast label={toast.message} isDleteBtn onClose={() => setToast(null)} style={{ zIndex: "9999999" }} error />
+        <Toast style={{ bottom: "5.5rem", zIndex: "9999999" }} label={toast.message} isDleteBtn onClose={() => setToast(null)} error />
       )}
       {toast && toast.state === "warning" && (
-        <Toast label={toast.message} isDleteBtn onClose={() => setToast(null)} style={{ zIndex: "9999999" }} warning />
+        <Toast style={{ bottom: "5.5rem", zIndex: "9999999" }} label={toast.message} isDleteBtn onClose={() => setToast(null)} warning />
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:672]: The intend for bottom style was to move the Toast a bit up.
Rest of the styling is already done within the component.

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js [hunk 3]
```diff
@@ -1,52 +1,157 @@
-import { FormComposerV2 } from "@egovernments/digit-ui-react-components";
+// import { ActionBar, FormComposerV2 } from "@egovernments/digit-ui-react-components";
+// import { ActionBar, Button, Stepper} from "@egovernments/digit-ui-components";
+import { ActionBar, Stepper, Toast } from "@egovernments/digit-ui-components";
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js:3]: Button in digit ui is not working well with action bar so i am using the button from react components

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js [hunk 4]
```diff
@@ -1,52 +1,157 @@
...
     if (!timeLineOptions || timeLineOptions.length === 0) return;
     setCurrentPage(timeLineOptions[0]);
   }, [timeLineOptions]);
 
+  // Effect to handle data completion validation and show toast
+  useEffect(() => {
+    if (checkDataCompletion === "false" || checkDataCompletion === "true" || checkDataCompletion === "valid") return;
+    if (checkDataCompletion === "invalid") setToast({ label: t("ERROR_DATA_INCOMPLETE"), error: true });
+    setCheckDataCompletion("false");
+  }, [checkDataCompletion]);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js:49]: I dont want  setCheckDataCompletion("false") to run if checkDataCompletion is either "false" "true" or "valid"

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js [hunk 5]
```diff
@@ -1,52 +1,157 @@
...
+    if (!microplanData) return;
+    Digit.SessionStorage.set("microplanData", microplanData);
+  }, [microplanData]);
+
+  // useEffect to store data in session storage
+  useEffect(() => {
+    const data = Digit.SessionStorage.get("microplanData");
+    setMicroplanData(data);
+    if (data && data.currentPage) setCurrentPage(data.currentPage);
+  }, []);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js:71]: one is to set the store the data in session storage and other one it to fetch the data on first render

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js [hunk 6]
```diff
@@ -1,52 +1,157 @@
...
+      <Stepper
+        type="stepper"
+        currentStep={currentPage?.id + 1}
+        customSteps={timeLineOptions.map((item) => t(item.name))}
+        direction="horizontal"
+        // onStepClick={() => {
+        //   setCheckDataCompletion("true");
+        //   setNavigationEvent("step");
+        // }}
+        onStepClick={onStepClick}
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/CreateMicroplan.js:104]: That is a needed functionality, just that it was disabled as all the screens are not yet made and the developers need a way to navigate between screens.

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/components/Nagivator.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    </div>
+  );
+};
+
+// Component to load custom component based on current page
+const LoadCustomComponent = (props) => {
+  if (props && !props.component) return null;
+  const secondaryProps = props.secondaryProps;
+  return <props.component {...secondaryProps} />;
+};
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/components/Nagivator.js:153]: No it is not defined within the navigator

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 8]
```diff
@@ -42,6 +50,46 @@ const Upload = ({ MicroplanName = "default", campaignType = "SMC" }) => {
   const [previewUploadedData, setPreviewUploadedData] = useState();
   const [uploadGuideLines, setUploadGuideLines] = useState();
 
+  // UseEffect for checking completeness of data before moveing to next section
+  useEffect(() => {
+    if (!fileDataList || checkDataCompletion !== "true" || !setCheckDataCompletion) return;
+    const valueList = fileDataList ? Object.values(fileDataList) : [];
+    if (valueList.length !== 0 && fileDataList.Microplanning_Population && fileDataList.Microplanning_Population.error === null) setCheckDataCompletion("valid");
+    else setCheckDataCompletion("invalid");
+  }, [checkDataCompletion]);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:59]: That is exactly what is happening in the current code

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 9]
```diff
@@ -13,8 +13,16 @@ import { SpatialDataPropertyMapping } from "../../components/resourceMapping";
 import shp from "shpjs";
 import { JsonPreviewInExcelForm } from "../../components/JsonPreviewInExcelForm";
 import { ButtonType1, ButtonType2, ModalHeading } from "../../components/ComonComponents";
-
-const Upload = ({ MicroplanName = "default", campaignType = "SMC" }) => {
+import { Toast } from "@egovernments/digit-ui-components";
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:16]: It is being used

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 10]
```diff
@@ -612,18 +665,18 @@ const Upload = ({ MicroplanName = "default", campaignType = "SMC" }) => {
...
       {loaderActivation && <LoaderWithGap text={"FILE_UPLOADING"} />}
-      {toast && toast.state === "success" && <Toast label={toast.message} onClose={() => setToast(null)} />}
+      {toast && toast.state === "success" && <Toast style={{ bottom: "5.5rem" }} label={toast.message} onClose={() => setToast(null)} />}
       {toast && toast.state === "error" && (
-        <Toast label={toast.message} isDleteBtn onClose={() => setToast(null)} style={{ zIndex: "9999999" }} error />
+        <Toast style={{ bottom: "5.5rem", zIndex: "9999999" }} label={toast.message} isDleteBtn onClose={() => setToast(null)} error />
       )}
       {toast && toast.state === "warning" && (
-        <Toast label={toast.message} isDleteBtn onClose={() => setToast(null)} style={{ zIndex: "9999999" }} warning />
+        <Toast style={{ bottom: "5.5rem", zIndex: "9999999" }} label={toast.message} isDleteBtn onClose={() => setToast(null)} warning />
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:679]: Rest of the Toast styling is done inside the component.

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 11]
```diff
@@ -682,18 +732,30 @@ const UploadComponents = ({ item, selected, uploadOptions, selectedFileType, sel
...
+    const [isHovered, setIsHovered] = useState(false);
+
+    const handleMouseEnter = () => {
+      setIsHovered(true);
+    };
+
+    const handleMouseLeave = () => {
+      setIsHovered(false);
+    };
+
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:744]: It is contributing.

**Review by nipunarora-eGov — APPROVED** [review]

**Pull request merged**